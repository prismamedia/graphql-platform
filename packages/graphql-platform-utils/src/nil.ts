export type Nil = undefined | null;

export function isNil(maybeNil: unknown): maybeNil is Nil {
  return maybeNil == null;
}

export type Nillable<T> = T | Nil;

export type NonNillable<T> = Exclude<T, Nil>;
