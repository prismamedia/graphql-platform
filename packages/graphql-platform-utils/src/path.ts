import { addPath as baseAddPath, type Path } from 'graphql/jsutils/Path.js';
import assert from 'node:assert/strict';
import type { InputType } from './input/type.js';
import { isPlainObject } from './plain-object.js';

export * from 'graphql/jsutils/Path.js';

export const isPath = (maybePath: unknown): maybePath is Path =>
  isPlainObject(maybePath) &&
  ['string', 'number'].includes(typeof maybePath['key']) &&
  (maybePath['prev'] === undefined || isPath(maybePath['prev']));

export const addPath = (
  prev: Path | undefined,
  key: string | number,
  type?: string | InputType,
): Path => baseAddPath(prev, key, type != null ? String(type) : undefined);

export function isPathDescendantOf(path: Path, maybeAncestor: Path): boolean {
  assert.notEqual(
    path,
    maybeAncestor,
    `Expects the "path" to be different of the "maybeAncestor"`,
  );

  return path.prev
    ? path.prev === maybeAncestor ||
        isPathDescendantOf(path.prev, maybeAncestor)
    : false;
}

export function arePathsEqual(a: Path, b: Path): boolean {
  return (
    a === b ||
    (a.key === b.key &&
      (a.prev && b.prev ? arePathsEqual(a.prev, b.prev) : !a.prev && !b.prev))
  );
}

export function isPathEqualOrDescendantOf(
  path: Path,
  maybeAncestor: Path,
): boolean {
  return path === maybeAncestor || isPathDescendantOf(path, maybeAncestor);
}

export function isPathAncestorOf(path: Path, maybeDescendant: Path): boolean {
  assert.notEqual(
    path,
    maybeDescendant,
    `Expects the "path" to be different of the "maybeDescendant"`,
  );

  return isPathAncestorOf(maybeDescendant, path);
}

export function getRelativePath(path: Path, ancestor: Path): Path {
  assert(
    isPathDescendantOf(path, ancestor),
    `Expects the "path" to be a descendant of the "ancestor"`,
  );

  return addPath(
    path.prev && path.prev !== ancestor
      ? getRelativePath(path.prev, ancestor)
      : undefined,
    path.key,
    path.typename,
  );
}

export const printPath = (path: Path, ancestor?: Path): string =>
  ancestor
    ? `.${printPath(getRelativePath(path, ancestor))}`
    : path.prev
      ? `${printPath(path.prev)}/${path.key}`
      : `/${path.key}`;
