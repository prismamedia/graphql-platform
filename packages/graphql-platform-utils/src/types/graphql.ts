export {
  FieldConfig as GraphQLFieldConfig,
  FieldConfigMap as GraphQLFieldConfigMap,
  FieldResolver as GraphQLFieldResolver,
  FieldResolverArgs as GraphQLFieldResolverArgs,
} from './graphql/field';
export { OperationType as GraphQLOperationType } from './graphql/operation-type';
