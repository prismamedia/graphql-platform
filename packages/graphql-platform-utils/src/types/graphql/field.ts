import {
  GraphQLFieldConfig as GraphQLFieldConfigBase,
  GraphQLResolveInfo,
} from 'graphql';
import { MaybePromise } from '../maybe-promise';

export type FieldResolverArgs<TSource = any, TContext = any, TArgs = any> = [
  TSource,
  TArgs,
  TContext,
  GraphQLResolveInfo,
];

/** Override the original "GraphQLFieldResolver" to be able to define the result */
export type FieldResolver<
  TSource = any,
  TContext = any,
  TArgs = any,
  TResult = any
> = (
  ...args: FieldResolverArgs<TSource, TContext, TArgs>
) => MaybePromise<TResult>;

/** Override the original "GraphQLFieldConfig" to be able to define the resolver's result */
export interface FieldConfig<
  TSource = any,
  TContext = any,
  TArgs = any,
  TResult = any
> extends Omit<GraphQLFieldConfigBase<TSource, TContext, TArgs>, 'resolve'> {
  resolve?: FieldResolver<TSource, TContext, TArgs, TResult>;
}

export interface FieldConfigMap<TSource = any, TContext = any, TArgs = any> {
  [key: string]: FieldConfig<TSource, TContext, TArgs>;
}
