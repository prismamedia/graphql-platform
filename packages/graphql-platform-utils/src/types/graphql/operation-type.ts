// cf: https://graphql.github.io/graphql-spec/draft/#OperationType
export enum OperationType {
  Mutation = 'mutation',
  Query = 'query',
  Subscription = 'subscription',
}
