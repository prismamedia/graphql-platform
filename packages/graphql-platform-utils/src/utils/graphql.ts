export * from './graphql/get-enum-type';
export * from './graphql/list-decorator';
export * from './graphql/non-null-decorator';
export * from './graphql/parse-resolver-info';
