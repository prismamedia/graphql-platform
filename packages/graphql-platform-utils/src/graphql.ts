import * as graphql from 'graphql';
import { isDeepStrictEqual } from 'node:util';
import { castToError, UnexpectedValueError } from './error.js';
import { indefinite } from './indefinite.js';
import { isNil, type Nillable } from './nil.js';
import { type Path } from './path.js';
import { isPlainObject, type PlainObject } from './plain-object.js';

export const parseGraphQLUntypedArgumentNodes = (
  argumentNodes: ReadonlyArray<graphql.ArgumentNode>,
  variables?: graphql.GraphQLResolveInfo['variableValues'],
) =>
  argumentNodes.reduce<PlainObject>((result, argument) => {
    const value = graphql.valueFromASTUntyped(argument.value, variables);

    return value !== undefined
      ? Object.assign(result, { [argument.name.value]: value })
      : result;
  }, Object.create(null));

export function isGraphQLResolveInfo(
  maybeGraphQLResolveInfo: unknown,
): maybeGraphQLResolveInfo is graphql.GraphQLResolveInfo {
  return (
    isPlainObject(maybeGraphQLResolveInfo) &&
    typeof maybeGraphQLResolveInfo.fieldName === 'string' &&
    Array.isArray(maybeGraphQLResolveInfo.fieldNodes) &&
    graphql.isOutputType(maybeGraphQLResolveInfo.returnType) &&
    graphql.isOutputType(maybeGraphQLResolveInfo.parentType)
  );
}

export function assertGraphQLResolveInfo(
  maybeGraphQLResolveInfo: unknown,
  path?: Path,
): asserts maybeGraphQLResolveInfo is graphql.GraphQLResolveInfo {
  if (!isGraphQLResolveInfo(maybeGraphQLResolveInfo)) {
    throw new UnexpectedValueError(
      `a GraphQLResolveInfo`,
      maybeGraphQLResolveInfo,
      { path },
    );
  }
}

export function isGraphQLASTNode<TKind extends graphql.Kind>(
  maybeGraphQLASTNode: unknown,
  kind: TKind,
): maybeGraphQLASTNode is graphql.ASTKindToNode[TKind] {
  return (
    isPlainObject(maybeGraphQLASTNode) && maybeGraphQLASTNode.kind === kind
  );
}

export function assertGraphQLASTNode<TKind extends graphql.Kind>(
  maybeGraphQLASTNode: unknown,
  kind: TKind,
  path?: Path,
): asserts maybeGraphQLASTNode is graphql.ASTKindToNode[TKind] {
  if (!isGraphQLASTNode(maybeGraphQLASTNode, kind)) {
    throw new UnexpectedValueError(`a GraphQL ${kind}`, maybeGraphQLASTNode, {
      path,
    });
  }
}

export function parseGraphQLScalarValue<TInternal>(
  type: graphql.GraphQLScalarType<TInternal, any>,
  maybeScalarValue: unknown,
  path?: Path,
): Nillable<TInternal> {
  if (isNil(maybeScalarValue)) {
    return maybeScalarValue;
  }

  try {
    return type.parseValue(maybeScalarValue);
  } catch (error) {
    throw new UnexpectedValueError(indefinite(type.name), maybeScalarValue, {
      path,
      cause: castToError(error),
    });
  }
}

export function parseGraphQLEnumValue<TInternal = any>(
  type: graphql.GraphQLEnumType,
  maybeEnumValue: unknown,
  path?: Path,
): Nillable<TInternal> {
  if (isNil(maybeEnumValue)) {
    return maybeEnumValue;
  }

  const enumValue = type
    .getValues()
    .find(({ value }) => isDeepStrictEqual(value, maybeEnumValue));

  if (!enumValue) {
    throw new UnexpectedValueError(
      `${indefinite(type.name)} (= a value among "${type
        .getValues()
        .map(({ value }) => value)
        .join(', ')}")`,
      maybeEnumValue,
      { path },
    );
  }

  return enumValue.value;
}

export function parseGraphQLLeafValue(
  type: graphql.GraphQLLeafType,
  maybeLeafValue: unknown,
  path?: Path,
): any {
  if (isNil(maybeLeafValue)) {
    return maybeLeafValue;
  }

  return graphql.isScalarType(type)
    ? parseGraphQLScalarValue(type, maybeLeafValue, path)
    : parseGraphQLEnumValue(type, maybeLeafValue, path);
}
