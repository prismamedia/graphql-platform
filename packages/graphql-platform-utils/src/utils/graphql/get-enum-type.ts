import { GraphQLEnumType, GraphQLEnumValueConfigMap } from 'graphql';
import { POJO } from '../../types/pojo';

export function getGraphQLEnumType(name: string, enumerable: POJO, useKeyAsValue: boolean = false): GraphQLEnumType {
  return new GraphQLEnumType({
    name,
    values: Object.entries(enumerable).reduce(
      (values: GraphQLEnumValueConfigMap, [key, value]) =>
        !Number.isNaN(Number.parseInt(key))
          ? values
          : Object.assign(values, { [key]: { value: useKeyAsValue ? key : value } }),
      Object.create(null),
    ),
  });
}
