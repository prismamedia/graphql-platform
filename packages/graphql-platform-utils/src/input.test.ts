import * as graphql from 'graphql';
import {
  Input,
  InputConfig,
  ListableInputType,
  nonNillableInputType,
  NonNullableInputType,
  NonOptionalInputType,
  ObjectInputType,
} from './input.js';
import { Nillable } from './nil.js';
import { addPath } from './path.js';

describe('Input', () => {
  const testPath = addPath(undefined, 'test');

  it.each<[config: InputConfig, error: string]>([
    [
      {
        name: 'InputWithInvalidStaticDefaultValue',
        type: graphql.GraphQLInt,
        defaultValue: 'a string',
      },
      `"InputWithInvalidStaticDefaultValue.defaultValue" - Expects to be valid against the type "Int" and the custom validation, got: 'a string'
└ Cause: "InputWithInvalidStaticDefaultValue" - Expects an "Int", got: 'a string'
  └ Cause: Int cannot represent non-integer value: "a string"`,
    ],
    [
      {
        name: 'InputWithInvalidThunkedDefaultValue',
        type: graphql.GraphQLInt,
        defaultValue: () => 'a string',
      },
      `"InputWithInvalidThunkedDefaultValue.defaultValue" - Expects to be valid against the type "Int" and the custom validation, got: 'a string'
└ Cause: "InputWithInvalidThunkedDefaultValue" - Expects an "Int", got: 'a string'
  └ Cause: Int cannot represent non-integer value: "a string"`,
    ],
    [
      {
        name: 'InputWithCustomValidationAndInvalidThunkedDefaultValue',
        type: graphql.GraphQLInt,
        assertValue: (value) => {
          if (value < 18) {
            throw new Error('Must be greater than 18');
          }
        },
        defaultValue: () => 16,
      },
      `"InputWithCustomValidationAndInvalidThunkedDefaultValue.defaultValue" - Expects to be valid against the type "Int" and the custom validation, got: 16
└ Cause: "InputWithCustomValidationAndInvalidThunkedDefaultValue" - Must be greater than 18`,
    ],
  ])('cannot have an invalid defaultValue', (config, error) => {
    expect(() => new Input(config)).toThrowError(error);
  });

  it.each<
    [
      input: Input,
      required: boolean,
      success?: [input: any, result?: any][],
      errors?: [input: any, error: any][],
    ]
  >([
    [
      new Input({ name: 'integer', type: graphql.GraphQLInt }),
      false,
      [
        [undefined, undefined],
        [null, null],
        [10, 10],
      ],
      [
        [{}, 'Expects an "Int", got: {}'],
        [true, 'Expects an "Int", got: true'],
        ['0', 'Expects an "Int", got: \'0\''],
      ],
    ],
    [
      new Input({ name: 'string', type: graphql.GraphQLString }),
      false,
      [
        [undefined, undefined],
        [null, null],
        ['my string', 'my string'],
      ],
      [
        [{}, 'Expects a "String", got: {}'],
        [true, 'Expects a "String", got: true'],
        [0, 'Expects a "String", got: 0'],
      ],
    ],
    [
      new Input({
        name: 'nonNillableString',
        type: nonNillableInputType(graphql.GraphQLString),
      }),
      true,
      undefined,
      [
        [undefined, 'Expects a non-undefined "String"'],
        [null, 'Expects a non-null "String"'],
      ],
    ],
    [
      new Input({
        name: 'nonNillableStringWithDefaultValue',
        type: nonNillableInputType(graphql.GraphQLString),
        defaultValue: 'My default value',
      }),
      false,
      [[undefined, 'My default value']],
      [[null, 'Expects a non-null "String"']],
    ],
    [
      new Input({
        name: 'nonNullableString',
        type: new NonNullableInputType(graphql.GraphQLString),
      }),
      false,
      [
        [undefined, undefined],
        ['my string', 'my string'],
      ],
      [[null, 'Expects a non-null "String"']],
    ],
    [
      new Input({
        name: 'nonOptionalString',
        type: new NonOptionalInputType(graphql.GraphQLString),
      }),
      true,
      [
        [null, null],
        ['my string', 'my string'],
      ],
      [[undefined, 'Expects a non-undefined "String"']],
    ],
    [
      new Input({
        name: 'nonOptionalObject',

        type: new NonOptionalInputType(
          new ObjectInputType({
            name: 'MyObject',
            fields: () => [
              new Input({
                name: 'username',
                type: graphql.GraphQLString,
              }),
            ],
          }),
        ),
      }),
      true,
      [
        [null, null],
        [{}, {}],
        [{ username: undefined }, {}],
        [{ username: null }, { username: null }],
        [{ username: 'yvann' }, { username: 'yvann' }],
      ],
      [[undefined, 'Expects a non-undefined "MyObject"']],
    ],
  ])('"%s" works', (input, required, success = [], errors = []) => {
    expect(input.isRequired()).toBe(required);

    if (input.isPublic()) {
      expect(input.getGraphQLConfig()).toBeDefined();
    }

    success.forEach(([value, result]) =>
      expect(input.parseValue(value)).toEqual(result),
    );

    errors.forEach(([value, error]) =>
      expect(() => input.parseValue(value)).toThrow(error),
    );

    expect(input.getGraphQLConfig()).toMatchSnapshot(input.name);
  });

  it('throws an Error on invalid value', () => {
    let objectInputType: ObjectInputType;

    objectInputType = new ObjectInputType({
      name: 'User',
      fields: () => [
        new Input({
          name: 'firstname',
          type: nonNillableInputType(graphql.GraphQLString),
        }),
        new Input({
          name: 'lastname',
          type: graphql.GraphQLString,
        }),
        new Input<Nillable<number>>({
          name: 'age',
          type: graphql.GraphQLInt,
          assertValue: (value) => {
            if (value < 18) {
              throw new Error('Must be greater than 18');
            }

            return value;
          },
        }),
        new Input({
          name: 'friends',
          type: new NonNullableInputType(
            new ListableInputType(nonNillableInputType(objectInputType)),
          ),
        }),
      ],
    });

    const invalidObject = {
      lastname: 'Boucher',
      friends: [
        { firstname: 'Maxime', friends: null },
        { lastname: 'Hubert', age: 17 },
        {
          firstname: 'Baptiste',
          friends: [
            { firstname: 'Maxime', friends: null },
            { firstname: 'Hubert', friends: 5 },
            null,
            { age: 10 },
          ],
        },
      ],
    };

    expect(objectInputType.validate()).toBeUndefined();

    expect(() => objectInputType.parseValue(invalidObject, testPath))
      .toThrowErrorMatchingInlineSnapshot(`
      ""test" - 2 errors:
      └ "firstname" - Expects a non-undefined "String"
      └ "friends" - 3 errors:
        └ "0.friends" - Expects a non-null "[User!]"
        └ "1" - 2 errors:
          └ "firstname" - Expects a non-undefined "String"
          └ "age" - Must be greater than 18
        └ "2.friends" - 4 errors:
          └ "0.friends" - Expects a non-null "[User!]"
          └ "1.friends" - Expects a plain-object, got: 5
          └ "2" - Expects a non-null "User"
          └ "3" - 2 errors:
            └ "firstname" - Expects a non-undefined "String"
            └ "age" - Must be greater than 18"
    `);

    expect(() =>
      objectInputType.parseValue(invalidObject, addPath(testPath, 'MyUser')),
    ).toThrowErrorMatchingInlineSnapshot(`
      ""test.MyUser" - 2 errors:
      └ "firstname" - Expects a non-undefined "String"
      └ "friends" - 3 errors:
        └ "0.friends" - Expects a non-null "[User!]"
        └ "1" - 2 errors:
          └ "firstname" - Expects a non-undefined "String"
          └ "age" - Must be greater than 18
        └ "2.friends" - 4 errors:
          └ "0.friends" - Expects a non-null "[User!]"
          └ "1.friends" - Expects a plain-object, got: 5
          └ "2" - Expects a non-null "User"
          └ "3" - 2 errors:
            └ "firstname" - Expects a non-undefined "String"
            └ "age" - Must be greater than 18"
    `);
  });
});
