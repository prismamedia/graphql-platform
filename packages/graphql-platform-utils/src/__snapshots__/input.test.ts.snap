// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input "integer" works: integer 1`] = `
Object {
  "type": "Int",
}
`;

exports[`Input "nonNillableString" works: nonNillableString 1`] = `
Object {
  "type": "String!",
}
`;

exports[`Input "nonNillableStringWithDefaultValue" works: nonNillableStringWithDefaultValue 1`] = `
Object {
  "defaultValue": "My default value",
  "type": "String!",
}
`;

exports[`Input "nonNullableString" works: nonNullableString 1`] = `
Object {
  "type": "String",
}
`;

exports[`Input "nonOptionalObject" works: nonOptionalObject 1`] = `
Object {
  "type": "MyObject",
}
`;

exports[`Input "nonOptionalString" works: nonOptionalString 1`] = `
Object {
  "type": "String",
}
`;

exports[`Input "string" works: string 1`] = `
Object {
  "type": "String",
}
`;
